### PR TITLE
Implemented PR_AUC.

### PR DESCRIPTION
**Pull request by vpatryshev**
Removed two obsolete methods in Functions (using Java7.Objects)
Implemented auc for arbitrary parameterized curves.

### CODE REVIEW

**Comment by ledell** [comment]
@vpatryshev Can you describe what this PR adds (which functions and what do they do / what do they return)?  At first glance, it looks like you've added two public functions, `pr_auc` and `AUC`, where `pr_auc` is the main user-facing method to extract and AUC is the worker function, or something to that effect.  

We should make sure this is somewhat consistent with the R/Py API (in Scala flavor).

@mmalohlava @vpatryshev In R, Py, we use the name, `auc()` as a method which extracts the AUC from a performance metrics object.  So I'd think we'd want something similar?

**Comment by mmalohlava** [comment]
1st iteration we just need a simple function which computes pr auc, which Avkash can send.

**Comment by vpatryshev** [comment]
@ledell The plan was to implement the feature request from PUBDEF 4099. The request was to enable producing PR AUC. We already have all the necessary data, so now that this function is available, anybody from Java or Scala can have it. If there's a demand for pr_auc from python (or R), I'll have to add a method similar to auc(). 
The method Functions.AUC just calculates an integral for a parameterized curve (specified by two functions). It would be nice if other code used it as well.

**Comment by ledell** [comment]
@vpatryshev Ah, ok, so this actually computes AUC, rather than retrieves it from a metrics object... so yeah, then the name doesn't matter.

**Comment by Avkash** [comment]
Good.

**Comment by vpatryshev** [comment]
Avkash, thanks!
What are the wellness criteria?

Thanks,
-Vlad

**Comment by mmalohlava** [comment]
> Please make sure the PRAUC is calculated well.

that's purpose of the review to verify that it is computed well

File: h2o-core/src/main/java/water/udf/fp/Functions.java [hunk 1]
```diff
@@ -81,19 +83,40 @@ public int hashCode() {
...
+      double x1 = x.apply(i);
+      double y1 = y.apply(i);
+      s += (y1+y0)*(x1-x0)*.5;
+
+      x0 = x1; y0 = y1;
+    }
 
-  public static boolean equal(Object x, Object y) {
-    return x == null ? y == null : x.equals(y);
+    return s;
```
Comment (mmalohlava) [inline h2o-core/src/main/java/water/udf/fp/Functions.java]: Do we need to rescale here with respect to `_n` and `_p`?

File: h2o-core/src/main/java/water/udf/fp/Functions.java [hunk 2]
```diff
@@ -81,19 +83,40 @@ public int hashCode() {
...
+   * that is, for a parametric curve specified by functions x and y,
+   * defined on integer domain [from, to], calculate the area
+   * between x[from], x[to], horizontal axis, and the curve.
+   * @param x x-component of the curve
+   * @param y y-component of the curve
+   * @param from min value of the curve range
+   * @param to max value of the curve range
+   * @return the area under curve, the result of integrating x*y' over [from,to].
+   */
+  public static double integrate(
```
Comment (mmalohlava) [inline h2o-core/src/main/java/water/udf/fp/Functions.java]: What about corner cases - `_fps[nbins-1] == 0` or `_tps[nbins-1] == 0` ?

File: h2o-core/src/main/java/water/udf/fp/Functions.java (hunk 1 above)
Comment (vpatryshev) [inline h2o-core/src/main/java/water/udf/fp/Functions.java]: Definitely not.

File: h2o-core/src/main/java/water/udf/fp/Functions.java (hunk 2 above)
Comment (vpatryshev) [inline h2o-core/src/main/java/water/udf/fp/Functions.java]: This check has been added in AUC2.

File: scripts/run.py [hunk 3]
```diff
@@ -388,13 +388,17 @@ def start(self):
                                           cwd=there)
             os.chdir(cwd)
         else:
-            self.child = subprocess.Popen(args=cmd,
-                                          stdout=f,
-                                          stderr=subprocess.STDOUT,
-                                          cwd=self.output_dir)
+            try:
```
Comment (mmalohlava) [inline scripts/run.py]: This is unrelated to PR AUC, please remove.

File: h2o-py/tests/testdir_algos/gbm/pyunit_PUBDEV_4090_grid_model_seeds.py [hunk 4]
```diff
@@ -72,7 +72,10 @@ def random_grid_model_seeds_PUBDEV_4090():
     model_len = min(len(air_grid1), len(air_grid2))
     correct_model_seeds = list(range(search_criteria_tune1["seed"], search_criteria_tune1["seed"]+model_len))
     assert model_seeds1[0:model_len]==model_seeds2[0:model_len], "Random gridsearch seed is not set correctly as model seeds..."
-    assert model_seeds1==correct_model_seeds, "Random gridsearch seed is not "
+    # this assert was failing; have to add more diagnostics
```
Comment (mmalohlava) [inline h2o-py/tests/testdir_algos/gbm/pyunit_PUBDEV_4090_grid_model_seeds.py]: This is unrelated to the PR AUC.

**Comment by stefan8888** [comment]
@vpatryshev It would be nice to have the pr_auc function available from python as well
